main: add a --version argument

supakeen opened the pull request:
Add `__version__` to `__init__.py` and an argument to show it.

This closes #992 by adding a `--version` argument.

```
user@desktopa osbuild € python3 -m osbuild --version
osbuild 58
```

and

```
user@desktopa osbuild € python3 -m osbuild          
usage: osbuild [-h] [--store DIRECTORY] [-l DIRECTORY] [--checkpoint ID]
               [--export ID] [--json] [--output-directory DIRECTORY]
               [--inspect] [--monitor NAME] [--monitor-fd FD]
               [--stage-timeout STAGE_TIMEOUT] [--version]
               MANIFEST
osbuild: error: the following arguments are required: MANIFEST
```

gicmo commented on osbuild/main_cli.py:
```diff
@@ -86,13 +87,25 @@ def parse_arguments(sys_argv):
                         help="File descriptor to be used for the monitor")
     parser.add_argument("--stage-timeout", type=int, default=None,
                         help="set the maximal time (in seconds) each stage is allowed to run")
+    parser.add_argument("--version", action="store_true",
+                        help="return the version of osbuild")
```
There is `parser.add_argument('--version', action='version', version='<the version>')`. The docs give the following example:
```
>>> import argparse
>>> parser = argparse.ArgumentParser(prog='PROG')
>>> parser.add_argument('--version', action='version', version='%(prog)s 2.0')
>>> parser.parse_args(['--version'])
PROG 2.0
```

supakeen replied:
I totally missed that existing, that solves the issue with `MANIFEST` as well.

gicmo commented on osbuild/__init__.py:
```diff
@@ -16,3 +16,10 @@
+try:
+    from pkg_resources import get_distribution
+
+    __version__ = get_distribution("osbuild").version
```
This probably does not work for non-installed `osbuild`, i.e. when running from the source dir.
Corner case: osbuild is installed but we are still running from the source dir. I wonder if we should care about this case.

supakeen replied:
Yea, in fact the exception of the distribution not being found should be caught and handled. The corner case needs some thought.

supakeen commented:
@gicmo It *is* possible to handle the version corner case of being inside the source repository but I don't know if that should be part of this PR :)

gicmo commented:
I wonder if we should use https://github.com/python-versioneer/python-versioneer

teg commented:
I think a related but possibly orthogonal feature may be to report the git sha. This would mean writing that to a file at rpm build time, though not sure we always have access to it.

I would love to be able to report the sha of all our tools, especially in the service where releases don't mean much.

I'm imagining a manifest generation result would contain the composer sha where the image definitions are from and similarly an image build result would contain the osbuild sha.

gicmo commented:
> I think a related but possibly orthogonal feature may be to report the git sha. This would mean writing that to a file at rpm build time, though not sure we always have access to it.

I think python-versioneer would solve this since it has `full-revisionid` via `get_versions`.

gicmo commented on setup.cfg:
```diff
@@ -10,3 +10,11 @@ max-statements=75
+versionfile_source = osbuild/_version.py
+versionfile_build = osbuild/_version.py
+tag_prefix = v
+parentdir_prefix = myproject-
```
`myproject-`, is that right?

supakeen replied:
No, it should be `osbuild-`. Done :)

supakeen commented:
@teg @gicmo Yea, using versioneer is great. I didn't want to do that initially because it touches a lot more code (setup.py, Makefile) and changes the rpmbuild output (kind-of).

However since you both seem positive I've updated this PR to use versioneer. See some examples:

```
user@desktopa osbuild € python3 -m osbuild --version
osbuild 58+2.g12317e0
user@desktopa osbuild € python3 setup.py --version
58+2.g12317e0
user@desktopa osbuild € ls -l rpmbuild/RPMS/noarch/
total 348
-rw-r--r--. 1 user user 109949 Jun 13 12:58 osbuild-58+2.g12317e0-1.fc36.noarch.rpm
-rw-r--r--. 1 user user  10863 Jun 13 12:58 osbuild-luks2-58+2.g12317e0-1.fc36.noarch.rpm
-rw-r--r--. 1 user user  10478 Jun 13 12:58 osbuild-lvm2-58+2.g12317e0-1.fc36.noarch.rpm
-rw-r--r--. 1 user user  20596 Jun 13 12:58 osbuild-ostree-58+2.g12317e0-1.fc36.noarch.rpm
-rw-r--r--. 1 user user  21490 Jun 13 12:58 osbuild-selinux-58+2.g12317e0-1.fc36.noarch.rpm
-rw-r--r--. 1 user user  19404 Jun 13 12:58 osbuild-tools-58+2.g12317e0-1.fc36.noarch.rpm
-rw-r--r--. 1 user user 150023 Jun 13 12:58 python3-osbuild-58+2.g12317e0-1.fc36.noarch.rpm
```

teg commented:
Du I understand correctly, that with PR we can know the git sha also for released RPMs? Or is that lost in the process? (Even if not it definitely looks worth it).

supakeen commented:
@teg Versioneer auto-determines the version number based on the tag (plus distance from the tag). For a package created at a tag it would 'just' be '58' or '59' without the git commit hash in there (perhaps possible to add it).

teg commented:
> @teg Versioneer auto-determines the version number based on the tag (plus distance from the tag). For a package created at a tag it would 'just' be '58' or '59' without the git commit hash in there (perhaps possible to add it).

Got it, thanks. I think when we are building rpms we may not have access to the git repo, only the contents of the checkout so wonder how that works.

supakeen commented:
> > @teg Versioneer auto-determines the version number based on the tag (plus distance from the tag). For a package created at a tag it would 'just' be '58' or '59' without the git commit hash in there (perhaps possible to add it).
> 
> Got it, thanks. I think when we are building rpms we may not have access to the git repo, only the contents of the checkout so wonder how that works.

Haha so, that's why there's about 500 force-pushes to this repo. I'm trying to get the RPM to build *correctly* on all flavors of CI we run.

This will be a long status update post with my unsolved problems :)

When using versioneer and making a dist (e.g.: `python setup.py sdist`) versioneer will replace `_version.py` with a compiled version that contains the version at the point of dist creation which is great, theoretically it should also write some correct placeholders on `git archive` but I haven't managed to get that to work.

So I replaced our `git archive` in the `Makefile` for `make rpm` with an sdist and added all necessary files to build this in our `MANIFEST.in` (files included in sdist). This makes `make rpm` work locally and includes the correct files end up inside the RPM:

```
user@desktopa noarch € rpm2cpio python3-osbuild-58+12.g6dcbdbf-1.fc36.noarch.rpm | cpio -idmf
1078 blocks
user@desktopa noarch € cat usr/lib/python3.10/site-packages/osbuild/_version.py 
...
 "version": "58+12.g6dcbdbf"
...
```

Now we get to the next problem, which is that we want to fill in the version in the `osbuild.spec` file, I want to have a single source of truth, so asking the package which version it is at is the best way there, it's compiled into the sdist and we can ask about it with:

```
Version:        %(python3 setup.py --version)
```

Which works fine when using rpmbuild but fails when using mockbuild as python3 is not available there:

```
Start: build phase for osbuild-58+12.g6dcbdbf-1.fc36.src.rpm
Start: build setup for osbuild-58+12.g6dcbdbf-1.fc36.src.rpm
sh: line 1: python3: command not found
error: line 4: Empty tag: Version:
Building target platforms: x86_64
```

And even when it *is* available (I added it for PackIt) we get a different failure in that `setup.py` is nowhere to be found (probably just another path):

```
2022-06-17 19:36:39.530 base_git.py       INFO   Using user-defined script for ActionName.get_current_version: [['python3', 'setup.py', '--version']]
2022-06-17 19:36:39.530 commands.py       DEBUG  Command: python3 setup.py --version
2022-06-17 19:36:39.722 logging.py        INFO   58+12.g6dcbdbf
...
2022-06-17 19:36:40.161 utils.py          ERROR  Preparation of the repository for creation of an SRPM failed: Failed to parse SPEC file: ["python3: can't open file '/workdir/setup.py': [Errno 2] No such file or directory", 'error: line 4: Empty tag: Version:']
Traceback (most recent call last):
  File "/usr/lib/python3.10/site-packages/rebasehelper/helpers/rpm_helper.py", line 182, in parse_spec
    result = rpm.spec(tmp.name, flags) if flags is not None else rpm.spec(tmp.name)
ValueError: can't parse specfile
```

And that's about where I'm stuck now. I want a single source of truth (the sdist is *always* the input for all RPM building and it contains the version correctly) but I also can't get that to work on all build systems.

Of course what you mentioned also applies. If a build system is doing a git archive of the repository and then building *in* that archive then tag information is not available and even sdist will do the wrong thing :(

supakeen commented:
Getting very close now, the GitLab CI and PackIt are both succesfully building the package (and I can build the package in all systems locally).

Local builds (either with rpmbuild *or* with mockbuild) succeed and write the correct (compiled) file into the RPM.

The PackIt system writes the (compiled) file but incorrectly, likely as it has no access to the git data initially. I see two options:

1. Give PackIt the git data.
2. Give PackIt an sdist that was generated from git.

supakeen commented:
It seems this marathon of CI is coming to an end! I figured out the last bits to correctly build images on PackIt which include the compiled `_version.py` file in the resultant (S)RPMs.

We might need to revisit the data that goes into the Version field in the spec files, it does not confirm to (for example) Fedora's guidelines: https://docs.fedoraproject.org/en-US/packaging-guidelines/Versioning/#_snapshots (replacing the + with a ^ might be enough in the RPM) but I'm unsure on the implications of not meeting the guidelines on these sort of transient package builds.

@gicmo, @teg or @ochosi maybe you have guidance on this? :)

teg commented:
> It seems this marathon of CI is coming to an end! I figured out the last bits to correctly build images on PackIt which include the compiled `_version.py` file in the resultant (S)RPMs.
> 
> We might need to revisit the data that goes into the Version field in the spec files, it does not confirm to (for example) Fedora's guidelines: https://docs.fedoraproject.org/en-US/packaging-guidelines/Versioning/#_snapshots (replacing the + with a ^ might be enough in the RPM) but I'm unsure on the implications of not meeting the guidelines on these sort of transient package builds. The 'normal' version is used for things built at a tag.
> 
> @gicmo, @teg or @ochosi maybe you have guidance on this? :)

I'm pretty sure that's ok, and that we only care about the guidelines for things we ship. Assuming it conforms to the rpm spec.

gicmo commented:
> > @gicmo, @teg or @ochosi maybe you have guidance on this? :)
> 
> I'm pretty sure that's ok, and that we only care about the guidelines for things we ship. Assuming it conforms to the rpm spec.

It might have sorting implications though.

gicmo commented:
I am very unsure about defining the version for the spec file from the outside, I have never seen that before. I think it would break the normal `fedpkg build`. I wonder if it also breaks auto-rebuilds that happen e.g. after branching and for new python versions.

supakeen commented:
@gicmo I'm starting to wonder if a non-versioneer approach might be better. We already have CI performing the post-release version-bump which could be extended to write `__version__` into `osbuild/__init__.py`.

We could slightly change it so it writes `N-dev` in the post-release version bump and we add a pre-release version bump to turn `N-dev` into `N`.

Commits built per commit are already reversioned by packit.

supakeen commented:
Let's stop chasing `versioneer` around, it doesn't mesh well with all the different ways we use the source to build packages. Instead this PR now uses the same bump-version logic as the rest uses by writing the version number to the `__init__.py`.

gicmo commented on Makefile:
```diff
@@ -301,3 +301,4 @@ bump-version:
 	sed -i "s|Version:\(\s*\)$(VERSION)|Version:\1$(NEXT_VERSION)|" osbuild.spec
 	sed -i "s|Release:\(\s*\)[[:digit:]]\+|Release:\11|" osbuild.spec
 	sed -i "s|version=\"$(VERSION)\"|version=\"$(NEXT_VERSION)\"|" setup.py
+	sed -i "s|__version__ = \"$(VERSION)\"|__version__ = \"$(NEXT_VERSION)\"|" osbuild/__init__.py
```
@ochosi do we need to adjust some bots for this as well?

ochosi reviewed (approved):
LGTM

supakeen commented:
The dependent PR is now merged as well. I think this one should be OK.
